ENH Add 'if_binary' option to drop argument of OneHotEncoder

rushabh-v opened the pull request:
Fixes #12502

rushabh-v committed:
add drop='if_binary' option

rushabh-v committed:
remove trailing spaces

rushabh-v committed:
add whatsnew

rushabh-v committed:
add isinstance of str condition

glemaitre commented on doc/whats_new/v0.23.rst:
```diff
@@ -114,6 +114,10 @@ Changelog
 
 - |Efficiency| :class:`preprocessing.OneHotEncoder` is now faster at
   transforming. :pr:`15762` by `Thomas Fan`_.
+- |Feature| argument `drop` of :class:`preprocessing.OneHotEncoder`
```
```suggestion

- |Feature| argument `drop` of :class:`preprocessing.OneHotEncoder`
```

glemaitre commented on doc/whats_new/v0.23.rst:
```diff
@@ -114,6 +114,10 @@ Changelog
   transforming. :pr:`15762` by `Thomas Fan`_.
+- |Feature| argument `drop` of :class:`preprocessing.OneHotEncoder`
+  will now accept value 'if_binary' and will drop the first category of
+  each binary features for `drop = 'if_binary'`. :pr:`#16245`
```
```suggestion
  each binary features for `drop='if_binary'`. :pr:`#16245`
```

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -293,7 +294,9 @@ def _validate_keywords(self):
         if self.drop is None:
             return None
-        elif (isinstance(self.drop, str) and self.drop == 'first'):
+        elif (isinstance(self.drop, str)
```
```suggestion
        elif isinstance(self.drop, str) and self.drop in ['first', 'if_binary']:
```

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+    ohe = OneHotEncoder(drop='if_binary')
+    ohe.fit(X)
+    result = ohe.transform(X).toarray()
+    assert_array_equal(expected, result)
```
```suggestion
    assert_allclose(expected, result)
```
The array is of dtype float

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+def test_one_hot_encoder_drop_equals_if_binary():
+    X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    expected = np.array([[1., 1., 0., 0.], [0., 0., 0., 1.], [0., 0., 1., 0.]])
+    ohe = OneHotEncoder(drop='if_binary')
```
```suggestion
    ohe = OneHotEncoder(drop='if_binary', sparse=False)
```

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+    expected = np.array([[1., 1., 0., 0.], [0., 0., 0., 1.], [0., 0., 1., 0.]])
+    ohe = OneHotEncoder(drop='if_binary')
+    ohe.fit(X)
+    result = ohe.transform(X).toarray()
```
```suggestion
    X_trans = ohe.fit_transform(X)
```

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+    X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    expected = np.array([[1., 1., 0., 0.], [0., 0., 0., 1.], [0., 0., 1., 0.]])
+    ohe = OneHotEncoder(drop='if_binary')
+    ohe.fit(X)
```
remove

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
             keep_cells = X_int != to_drop
+
+            if isinstance(self.drop, str) and self.drop == 'if_binary':
+                keep_cells = keep_cells.T
```
Why do you transpose here?

glemaitre replied:
We could just iterate the same way directly on the column?

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
+            if isinstance(self.drop, str) and self.drop == 'if_binary':
+                keep_cells = keep_cells.T
+                X_int = X_int.T
+                for i in range(n_features):
```
```suggestion
                for col in range(n_features):
```

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
+                keep_cells = keep_cells.T
+                X_int = X_int.T
+                for i in range(n_features):
+                    if len(self.categories_[i]) > 2:
```
```suggestion
                    if len(self.categories_[col]) > 2:
```

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
+                X_int = X_int.T
+                for i in range(n_features):
+                    if len(self.categories_[i]) > 2:
+                        keep_cells[i] = True
```
```suggestion
                        keep_cells[:, col] = True
```

jnothman replied:
If keep_cells is not transposed, you should be able to use: `keep_cells[:, i] = True`

glemaitre commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
+                    else:
+                        X_int[i][X_int[i] > 0] -= 1
+                keep_cells = keep_cells.T
+                X_int = X_int.T
```
No need to transpose

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
 
 
+def test_one_hot_encoder_drop_equals_if_binary():
+    X = [['Male', 1], ['Female', 3], ['Female', 2]]
```
You can add three additional columns (one binary and one mutli-category and again one binary).

We can ensure that there is nothing tricky with the ordering

rushabh-v replied:
Made all the changes

glemaitre reviewed (dismissed):
Thanks. It looks pretty good,

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
             keep_cells = X_int != to_drop
+
+            if isinstance(self.drop, str) and self.drop == 'if_binary':
+                keep_cells = keep_cells.T
```
I think you can avoid these transpositions.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +401,25 @@ def transform(self, X):
+                    if len(self.categories_[i]) > 2:
+                        keep_cells[i] = True
+                    else:
+                        X_int[i][X_int[i] > 0] -= 1
```
If X_int was not transposed, `X_int[X_int[i] > 0, i] -= 1` should work.

rushabh-v replied:
That raises this error `boolean index did not match indexed array along dimension 0; dimension is 3 but corresponding boolean dimension is 5`.
Using `X_int[:, col][X_int[:, col] > 0] -= 1` instead.

glemaitre replied:
you need to put the masking in the first indexing `[..., ...]`

jnothman commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,15 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
 
+def test_one_hot_encoder_drop_equals_if_binary():
+    X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    expected = np.array([[1., 1., 0., 0.], [0., 0., 0., 1.], [0., 0., 1., 0.]])
```
lay this out over three lines so that it is easier to see what's going on.

rushabh-v replied:
Made all the changes

rushabh-v committed:
make requested changes

rushabh-v committed:
solve linting errors

rth commented:
A bit related to categories="dtype" in https://github.com/scikit-learn/scikit-learn/pull/15396 I could imagine `if_binary=True` to be the only useful option long term. To keep backward compatibility constraints,  one possibility could be to implement `if_binary='warn'` and raise a deprecation warning indicating that the behaviour will change from `if_binary=False` to `True` by default in the future. Then some versions later remove `if_binary` altogether.

The alternative could be to make this breaking change in 1.0. WDYT? cc @thomasjpfan

rushabh-v commented:
>if_binary=True to be the only useful option long term.

Do you mean `drop='if_binary'`? @rth

rth commented:
> Do you mean drop='if_binary'

Yes, never mind my comment for this PR, particularly since this doesn't add a new parameter. The comment was more about how to make this the default in the long term.

amueller commented:
@thomasjpfan would be great if you could have a look ;)

rth reviewed (commented):
Thanks for the PR @rushabh-v !

So this currently drops the first collumn for any input with 2 classes, not just binary,
```py
>>> OneHotEncoder(drop=None).fit_transform([['a'], ['b'], ['a']]).A
array([[1., 0.],
       [0., 1.],
       [1., 0.]])
>>> OneHotEncoder(drop='if_binary').fit_transform([['a'], ['b'], ['a']]).A
array([[0.],
       [1.],
       [0.]])
```
I would expect to this only take effect for actually binary column, not any input with 2 classes. You could add it as a counter example in unit tests.

Enabling it if the drop='if_binary' the column is of dtype bool, might be easier. For int/float I suppose one could apply it if np.unique(X[col]) is [0, 1], but it's a bit more expensive.  Let's wait for input from other reviewers, before making changes.

glemaitre commented:
> I would expect to this only take effect for actually binary column, not any input with 2 classes. You could add it as a counter example in unit tests.

@rth how would you do to tag a feature as "binary"? I would somehow assume that 2 categories features should be close to a binary feature, isn't it?

jnothman commented:
I think here we are assuming that two categories means binary. The point of
this is that as long as there are only two categories, encoding both is
merely redundant.

rth commented:
> how would you do to tag a feature as "binary"? I would somehow assume that 2 categories features should be close to a binary feature, isn't it?
> 
> I think here we are assuming that two categories means binary.

OK that also works. Though, I think binary features has a very specific meaning, as either a column with binary dtype or a column with [0, 1]. A column with ["yes", "no"] or ["true", "false"], can be considered binary, but IMO it's up to the initial preprocessing to convert it to binary (also I have rarely seen this in practice). It might be best to avoid that vocabulary in this PR. Say if you have a column with countries `['JP', 'DE']` that's definitely not a binary feature, it can suggest that other categories are possible just not present in the data.

It might be OK to keep the name `drop="if_binary"` but document it without using "binary feature" term,
> - 'if_binary' : drop the first category for features that have 1 or 2 categories.

(since this PR also drops columns with 1 features I think). WDYT?

rushabh-v commented:
>since this PR also drops columns with 1 features I think

No, it doesn't drop columns with 1 category.

rushabh-v commented:
>since this PR also drops columns with 1 features I think

No, it doesn't the columns with 1 category.

rushabh-v committed:
Merge branch 'master' into ohe-binary

rth commented:
> No, it doesn't drop columns with 1 category.

Thanks for the confirmation. It is currently behaving strangely in that case though which probably needs fixing,
```py
>>> OneHotEncoder().fit_transform([['a'], ['a'], ['a']]).A
array([[1.],
       [1.],
       [1.]])
>>> OneHotEncoder(drop='if_binary').fit_transform([['a'], ['a'], ['a']]).A
array([[0.],
       [0.],
       [0.]])
```

In general just to clarify, implementation wise is this expected to behave identically to `drop='first'` but only apply with 2 classes (in which case we might as well just reuse the `drop='first'` implementation conditionally), or is there additional logic added in.

rushabh-v commented:
In every case when there is a binary feature, `drop='if_binary'` will work the same as `drop='first'`. But for `drop=None` the implementation is different. And that's why in the above example outputs for `drop='if_binary'` and `drop=None` are different. 

But IMO, for features with one category `drop=None` should also return all 0s.

rth commented:
> In every case when there is a binary feature, drop='if_binary' will work the same as drop='first'. But for drop=None the implementation is different. And that's why in the above example outputs for drop='if_binary' and drop=None are different.

The above example with `drop='first'` returns an empty array which more what I would expect,
```py
>>> OneHotEncoder(drop='first').fit_transform([['a'], ['a'], ['a']]).A
array([], shape=(3, 0), dtype=float64)
```

In terms of implementation I think we might want to change the signature of `_compute_drop_idx` to take X as input and make the computation there (r otherwise rethink where we compute `self.drop_idx_`, as currently it is not correct in some cases,
```py
>>> est = OneHotEncoder(drop='if_binary')
>>> est.fit(np.atleast_2d(['a', 'b', 'c']))
OneHotEncoder(categories='auto', drop='if_binary',
              dtype=<class 'numpy.float64'>, handle_unknown='error',
              sparse=True)
>>> est.drop_idx_
array([0, 0, 0])
```
In general we should compute `drop_idx_` then apply them, instead of adding custom logic after it is computed. I have not looked at the OHE code enough to say what would be the fastest to get there.

rushabh-v commented:
`drop_idx_` contains index of the category to be dropped from each feature. What should we put there in case of non-binary features then?

rushabh-v commented:
And even if we change the logic of `_compute_drop_idx`, then I think the logic to drop would become more expensive in case of `drop='if_binary'`. Can't we correct `drop_idx_` by putting some extra conditions or something?

rth commented:
> And even if we change the logic of _compute_drop_idx, then I think the logic to drop would become more expensive in case of drop='if_binary'. Can't we correct drop_idx_ by putting some extra conditions or something?

Yes, it may have some additional cost but basically the logic for drop should be,
 - in fit compute the columns to drop and store them in `drop_idx_`
 - in transform drop according to `drop_idx_`

so we cannot fix drop_idx_ in transform, as the determination of "binary" features need to happen based on train data in fit.

rushabh-v commented:
What I meant was that we can edit the `_compute_drop_idx` so that `drop_idx_` would be correct.

And  in case of `drop=if_binary` instead of using `drop_idx_` we can take a temporary variable `drop_bin = np.zeros(len(self.categories_), dtype=np.int_)` inside transform. And use that for all the computations.

correct me if I am wrong!

rth commented:
> What I meant was that we can edit the _compute_drop_idx so that drop_idx_ would be correct every time.

Yes, that is what I mean.

Something like: in fit,
```py
self.drop_idx_ = self._compute_drop_idx(self.categories)
```
with 
```py
def _compute_drop_idx(self, categories)
    ...
    elif (isinstance(self.drop, str) and self.drop == 'first'):
        return np.array([0 if len(col_cats) > 2 else None
                         for col_cats in categories], dtype=object)
```
and then in `transform` mostly what we have now with a check for `None` in `drop_idx_[i]`. Though that would mean that drop_idx_ can be an array of objects, while previously we always had an array of ints which might need some adjustment and documentation. Or we can use an int -1 (or some other negative value) to indicate that a column should not be dropped but that's also suboptimal.

rushabh-v commented:
So right now I am considering implementing it with object array just to check if it works fine or not. then we will change it if required.

rushabh-v committed:
use object np arrays for drop_idx_

rushabh-v committed:
Merge branch 'ohe-binary' of https://github.com/rushabh-v/scikit-learn into ohe-binary

rushabh-v committed:
fix linting errors

rushabh-v commented:
Every problem mentioned above got fixed with the latest changes.
```python
>>> OneHotEncoder().fit_transform([['a'], ['a'], ['a']]).A
array([[1.],
       [1.],
       [1.]])
>>> OneHotEncoder(drop='if_binary').fit_transform([['a'], ['a'], ['a']]).A
array([[0.],
       [0.],
       [0.]])
```

Even this also got fixed with this. The current behaviore is,
```python
>>> OneHotEncoder().fit_transform([['a'], ['a'], ['a']]).A
array([[1.],
       [1.],
       [1.]])
>>> OneHotEncoder(drop='if_binary').fit_transform([['a'], ['a'], ['a']]).A
array([[1.],
       [1.],
       [1.]])
```
I have added appended above case in the test and added a test for drop_idx_.

And it's even a bit more efficient :) because it will not have to compute `len(self.cetagory_[col])` in every iteration now.

rushabh-v committed:
add error msg as else part

rushabh-v commented:
are the object arrays okay or we need to switch to ints?

rth commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +408,21 @@ def transform(self, X):
+
+            else:
+                X_int[X_int > to_drop] -= 1
+                n_values = [len(cats) - 1 for cats in self.categories_]
```
So you are saying that we can't do a common implementation here that wouldn't use the `self.drop == 'if_binary'` information and only based on the previously computed drop_idx_?  If it's possible to do that, it would be ideal I think.

rushabh-v replied:
Yes, when I thought a bit more, I realized that drop_idx can be used for all the three cases without any extra tweaks.

rushabh-v replied:
Made all the changes

rushabh-v replied:
>So you are saying that we can't do a common implementation here that wouldn't use the self.drop == 'if_binary' information and only based on the previously computed drop_idx_? If it's possible to do that, it would be ideal I think.

Oh, I'm sorry, I misunderstood the question here. But I think no, we can't do this without putting that condition but in both the cases, drop_idx_ is used directly without any extra tweaks.

rth reviewed (commented):
> are the object arrays okay or we need to switch to ints?

I would prefer objects dtype, but let's see what other reviewers think. A few more comments below. Thanks!

rushabh-v committed:
make requested changes

rushabh-v committed:
solve linting error

rushabh-v commented:
There is a strange `urllib.error.URLError: <urlopen error [Errno 104] Connection reset by peer>
`.  see, https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=12870&view=logs&j=78a0bf4f-79e5-5387-94ec-13e67d216d6e&t=f1857171-4a53-55c7-3ab5-90acfe091baa&l=1047

rushabh-v committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into ohe-binary

thomasjpfan commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -398,9 +409,21 @@ def transform(self, X):
             # categories that occur after them to avoid an empty column.
 
             keep_cells = X_int != to_drop
+
```
For this `if self.drop is not None` block, this may be cleaner:

```py
        if self.drop is not None:
            to_drop = self.drop_idx_.reshape(1, -1).copy()

            keep_cells = X_int != to_drop
            n_values = []
            for i, cats in enumerate(self.categories_):
                n_cats = len(cats)
                if to_drop[0, i] is None:
                    # set to cardinality to not drop from X_int
                    to_drop[0, i] = n_cats
                    n_values.append(n_cats)
                else:  # dropped
                    n_values.append(n_cats - 1)

            X_int[X_int > to_drop] -= 1
            X_mask &= keep_cells
        else:
            n_values = [len(cats) for cats in self.categories_]

```

thomasjpfan reviewed (commented):
I have a slight preference to using `-1` over `None` for `drop_idx`.

The documentation of `drop_idx_` needs to be updated regarding this new behavior.

rushabh-v commented:
What should I finally use then, `None` or `-1`?

rth commented:
Let's follow @thomasjpfan's suggestion and use -1.

rushabh-v committed:
make requested changes

rth commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,51 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+                      [0., 1.],
+                      [0., 1.]]),
+
+            np.array([0, -1]),
```
We don't need to parametrize, as it's not very readable. We can just write these 3 cases one after the other in the test.

rushabh-v replied:
done and all greens

rth reviewed (approved):
LGTM, aside from one minor comment.

rushabh-v committed:
remove parametrization in test

rushabh-v committed:
fix linting

rushabh-v committed:
fix linting

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -195,6 +195,8 @@ class OneHotEncoder(_BaseEncoder):
         - 'first' : drop the first category in each feature. If only one
           category is present, the feature will be dropped entirely.
+        - 'if_binary' : drop the first category in each feature with two
+          categories.
```
Maybe add 


Features with 1 or more than 2 categories are left intact

NicolasHug commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -398,6 +398,47 @@ def test_one_hot_encoder_feature_names_drop(drop, expected_names):
+    expected = np.array([[1., 0., 0., 1.],
+                         [0., 1., 0., 0.],
+                         [0., 0., 1., 1.]])
+    expected_drop_idx = np.array([-1, 0])
```
It'd be useful to add short comments describing what you're testing.

e.g.

"Canonical case"

"with only one cat, the behaviour is equivalent to drop=None"

NicolasHug reviewed (commented):
A few comments but looks good. Thanks for the work @rushabh-v

rushabh-v commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -293,8 +297,17 @@ def _validate_keywords(self):
+                                for cats in self.categories_], dtype=np.int_)
+            else:
+                msg = ("Wrong input for parameter `drop`. Expected "
+                       "'first', None or array of objects, got {}")
```
>But isn't this message duplicated? I see the same below. It's better to have the failure logic in one single place

Doing that would increase the number of conditions to be checked.

rushabh-v replied:
Should I put them back in one place or not?

NicolasHug replied:
Well after double checking, the last `elif` is equivalent to an else:

```py

        elif isinstance(self.drop, str):
  			# ...

        elif not isinstance(self.drop, str):

```

So you can remove the `else` below, and make it replace the last elif. No more duplication.


The error message still needs to be updated though

rushabh-v committed:
make requested changes

NicolasHug commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -323,7 +343,7 @@ def _compute_drop_idx(self):
                              zip(self.drop, self.categories_)], dtype=np.int_)
         else:
```
I this this else be removed and you can replace the last elif with an else

rushabh-v replied:
Done

rushabh-v committed:
fix linting

rushabh-v committed:
make requested change

rushabh-v committed:
remove trailing whitespacce

NicolasHug reviewed (approved):
LGTM, thanks @rushabh-v !

NicolasHug commented:
I'm still gettig a 500 from codecov, not sure what's going on. Not merging just yet in case @thomasjpfan wants to have a look?

thomasjpfan reviewed (commented):
`inverse_transform` needs to be updated and tested with the new `'if_binary'` option.

jnothman commented:
Another small decision: if we look across the codebase, do we tend to use underscores, spaces, hyphens or other separators for words in param values? Should it be `'if_binary`' or `'if binary'` or `'ifbinary'`??

rth commented:
> Should it be 'if_binary' or 'if binary' or 'ifbinary'??

+1 for 'if_binary'

rushabh-v committed:
add if_binary to inverse_transform and add it's test

rushabh-v committed:
fix linting

rushabh-v committed:
fix inverse_transform

rushabh-v commented:
@thomasjpfan I have updated the inverse_transform and added the test, can you review, please?

thomasjpfan commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -265,6 +265,16 @@ def test_one_hot_encoder_inverse(sparse_, drop):
 
 
+def test_one_hot_encoder_inverse_if_binary():
+    X = [['Male', 1],
```
We can make the input a numpy array:

```py
    X = np.array([['Male', 1],
                  ['Female', 3],
                  ['Female', 2]])
```

and then

```py
assert_array_equal(ohe.inverse_transform(X_tr), X)
```

rushabh-v replied:
done

thomasjpfan reviewed (commented):
I am happy with `if_binary` as the keyword option.

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -448,8 +474,13 @@ def inverse_transform(self, X):
         else:
-            n_transformed_features = sum(len(cats) - 1
-                                         for cats in self.categories_)
+            if isinstance(self.drop, str) and self.drop == 'if_binary':
```
Please use elif if it is appropriate here

rushabh-v replied:
done

rushabh-v committed:
make requested changes

rushabh-v commented:
why it says `Cmd.exe exited with code '1'.`?

thomasjpfan commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -265,6 +265,15 @@ def test_one_hot_encoder_inverse(sparse_, drop):
+def test_one_hot_encoder_inverse_if_binary():
+    X = np.array([['Male', 1],
+                  ['Female', 3],
+                  ['Female', 2]])
```
Try:

```suggestion
                  ['Female', 2]], dtype=object)
```

Very interesting

rushabh-v committed:
Add dtype=object

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

rushabh-v commented:
All greens now!

rth commented:
Thanks @rushabh-v and all reviewers!
